File: Python/Two Pointers/M-1052. Grumpy Bookstore Owner.py
class Solution:
    def maxSatisfied(
        self, customers: List[int], grumpy: List[int], minutes: int
    ) -> int:
        init_satisfied = 0
        for i in range(len(customers)):
            if grumpy[i] == 1:
                continue
            init_satisfied += customers[i]
            customers[i] = 0

        current, max_satisfied = 0, 0
        l, r = 0, 0
        while r < len(customers):
            current += customers[r]
            if r - l + 1 >= minutes:
                max_satisfied = max(max_satisfied, current)
                current -= customers[l]
                l += 1
            r += 1
        return init_satisfied + max_satisfied

File: Javascript/Two Pointers/M-1052. Grumpy Bookstore Owner.ts
function maxSatisfied(
  customers: number[],
  grumpy: number[],
  minutes: number
): number {
  let initSatisfied = 0;
  for (let i = 0; i < customers.length; i++) {
    if (grumpy[i] === 0) {
      initSatisfied += customers[i];
      customers[i] = 0;
    }
  }

  let l = 0;
  let r = 0;
  let current = 0;
  let max = 0;
  while (r < customers.length) {
    current += customers[r];
    if (r - l + 1 === minutes) {
      max = Math.max(max, current);
      current -= customers[l];
      l++;
    }
    r++;
  }

  return initSatisfied + max;
}

const testData = [
  {
    customers: [1, 0, 1, 2, 1, 1, 7, 5],
    grumpy: [0, 1, 0, 1, 0, 1, 0, 1],
    minutes: 3,
    ans: 16,
  },
  {
    customers: [1],
    grumpy: [0],
    minutes: 1,
    ans: 1,
  },
];

for (const data of testData) {
  console.log(
    data.ans === maxSatisfied(data.customers, data.grumpy, data.minutes)
  );
}
